### DISCUSSION TITLE
Make the new, future no-profile CLI for use in shebang lines implement other CLI fixes too

### DISCUSSION THREAD

**Discussion by mklement0** [PowerShell/PowerShell#23872] [Enhancements]
<sup>Follow-up from a conversation with @SteveL-MSFT in #7989.</sup>

PowerShell's existing CLI has a number of problems, relating to:

* (a) unconditionally dot-sourcing `$PROFILE`, making for unpredictable execution environments - see #992
* (b) parameter syntax and parsing, both in general and with respect to fitting into the Unix world - see #3743
  * _Update_: One of the problems in the linked post has been mitigated in the _existing_ CLI in v7.5 (since preview 5), albeit not in a syntax-compatible way, because a _new parameter_, [`-CommandWithArgs`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Pwsh?view=powershell-7.5#-commandwithargs---cwa) was introduced that allows treating the first subsequent argument as a _script_ to which all the remaining arguments are passed as arguments.
* (c) unhelpful mapping of PowerShell's output streams to the outside world's standard streams - see #7989
* (d) the unhelpful pseudo-interactive behavior of `-File -` that contravenes expectations with respect to providing code via stdin - see #3223 

The existing CLI cannot be fixed so as not to break backward compatibility, but [it was agreed that (a) would be fixed by way of a _new, separate executable/symlink_](https://github.com/PowerShell/PowerShell/issues/992#issuecomment-384455760), notably for use in shebang lines of stand-alone shell scripts on Unix-like platforms.

More broadly, however, this new CLI is an opportunity to fix (b) and (c) as well, which would go a long way toward endearing PowerShell to CLI-savvy Unix users.

Essentially, it would provide a new CLI that is a well-behaved cross-platform citizen with sane defaults. 
(As such the profile-loading behavior should mimic Bash's - only load `$PROFILE` in _interactive_ shells by default).

Note: The new behavior needn't be a separate _binary_: The behavior variation could be triggered by invocation via _symlink_ whose name indicates the need to apply alternate behavior, at least on Unix-like platforms. (Invoking .NET Core executables via symlinks on Windows is not yet supported, but will hopefully come in .NET Core 3.0 - see https://github.com/dotnet/core-setup/issues/4080)

Given the then-broader scope, the originally proposed name `pwsh-np` needs revising, however.

Environment data
----------------

Written as of:

```powershell
PowerShell Core 6.1.0
```

**Comment by SteveL-MSFT** [comment]
Whatever name we choose, we also need to consider `pwsh-np-preview`

**Comment by mklement0** [comment]
I don't understand, @SteveL-MSFT: why `*-preview`?

**Comment by SteveL-MSFT** [comment]
@mklement0 we want to maintain a consistent separation of the stable and preview releases so I would think we'd have `pwsh`, `pwsh-preview` like we have today, along with `pwsh-np` and `pwsh-np-preview`

**Comment by mklement0** [comment]
Ah, of course, thanks @SteveL-MSFT.

Any thoughts on a suitable new name that reflects the scope and intent of the changes?

`psh`, perhaps? (OK, just kidding).

Naming is tricky, because the name should be kept short, so I fear it'll come down to documenting things well.

`pwshc`? (with the `c` standing for (a proper) CLI).

**Comment by tbakerweb** [comment]
Hey @SteveL-MSFT,  It isn't clear in this issue, could you please specify if/when/where the Streams output to the associated File Descriptors from #7989 will show up.  I'm very interested in this feature and was quite sad when I realized it didn't already exist :(.

Thanks for your contributions!

**Comment by sdwheeler** [comment]
@mklement0 Thanks for collecting these issues in one meta-issue. All of the linked issue have been closed but there is room for more discussion. The WG reviewed this and decided it would be better served by moving this issue to a GitHub Discussion where the community can help us with requirements and decisions for a future solution.

**Comment by rhubarb-geek-nz** [comment]
Given this discussion was started six years ago, I think the future has been and gone.

May I suggest that the current process is not working for this component?

I propose that people build examples/proofs of concept completely independently of this project/repository. There is no restriction on anyone writing a command line program that uses the SDK. For example I have [one that meets my needs](https://github.com/rhubarb-geek-nz/powershell-executive).

**Reply by mklement0** [reply]
Generally speaking, it is a logical fallacy to dismiss proposals based on the lack of activity around them, as opposed to dismissing them based on their (lack of) inherent merit.

It's fine to suggest pragmatic alternatives, but doing so doesn't warrant championing them as the only way forward.

If you see value in the proposal at hand, express it as such.

**Reply by rhubarb-geek-nz** [reply]
I am not dismissing the proposal, I am offering an opinion as to a way forward as I thought that was acceptable given it was moved to discussions.

**Reply by mklement0** [reply]
The only implied stance you've taken with respect to the proposal is "since no one's paying attention, let's move on" - my point is that this is a false dichotomy.

**Reply by rhubarb-geek-nz** [reply]
> The only implied stance you've taken with respect to the proposal is "since no one's paying attention, let's move on" 

No, my position is so lets do something about it. Get some prototypes, prove the concepts work, prove they don't break scp etc.
